src/encdatavault_fmt_plug.c

```diff
@@ -0,0 +1,407 @@
+
+	AES_set_encrypt_key(key, ENC_KEY_SIZE * 8, &aes_key);
+	len = len >> 4;
+	counter = counter << 56;
+
+	for (i = 0; i < len; i++) {
+		tmp_iv[0] = ((uint64_t *) ivs[0])[0];
+		tmp_iv[1] = counter;
+		AES_encrypt((unsigned char *)tmp_iv, out, &aes_key);
+		enc_xor_block((uint64_t *) out, (uint64_t *) in);
+
+		for (j = 1; j < nb_keys; j++) {
+			tmp_iv[0] = ((uint64_t *)(ivs[j]))[0];
+			tmp_iv[1] = counter;
+			AES_encrypt((unsigned char *)tmp_iv, (unsigned char *)tmp_iv,
+			            &aes_key);
+			enc_xor_block((uint64_t *) out, tmp_iv);
+		}
+		// Increment counter, only 255 block encryption is supported.
+		counter += 0x0100000000000000;
+		out += AES_BLOCK_SIZE;
+		in += AES_BLOCK_SIZE;
+	}
+}
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+	saved_key = mem_calloc_align(sizeof(*saved_key),
+	                             self->params.max_keys_per_crypt, MEM_ALIGN_WORD);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // algorithm id
+		goto err;
+	if (!isdec(p))
+		goto err;
+	if (atoi(p) > 4 || atoi(p) < 1) {
+		fprintf(stderr, "Warning: algorithm id %d not supported!\n", atoi(p));
+		goto err;
+	}
+	if ((p = strtokm(NULL, "$")) == NULL)   // Nonce
+		goto err;
+	if (hexlenl(p, &extra) != ENC_IV_PREFIX_SIZE * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // Encrypted header
+		goto err;
+	if (hexlenl(p, &extra) != ENC_SIG_SIZE * 2 || extra)
+		goto err;
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs = { 0 };
+	int i;
+	char *p = ciphertext, *ctcopy, *keeptr;
+
+	memset(&cs, 0, sizeof(cs));
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	cs.version = atoi(p);
+	p = strtokm(NULL, "$");
+	cs.algo_id = atoi(p);
+	p = strtokm(NULL, "$");
+	for (i = 0; i < ENC_IV_PREFIX_SIZE; i++)
+		cs.iv[i] =
+		    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i +
+		            1])];
+
+	// Init AES CTR counter
+	cs.iv[ENC_IV_SIZE - 1] = 1;
+
+	p = strtokm(NULL, "$");
+	for (i = 0; i < ENC_SIG_SIZE; i++)
+		cs.encrypted_data[i + 4] =
+		    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i +
+		            1])];
+
+	// Keychain for version 3
+	if (3 == cs.version) {
```

Comment:
Our code style would put this as `if (cs.version == 3) {`. Same for line 249.

Reply:
Corrected